File: rps_backend/utils/commit.py
import json
import secrets
from pathlib import Path

from web3 import Web3

from rps_backend.utils.config import LOCAL_SECRETS_DIR

MOVE_TO_INT = {
    "rock": 0,
    "paper": 1,
    "scissors": 2,
}

INT_TO_MOVE = {
    0: "rock",
    1: "paper",
    2: "scissors",
}


def normalize_move(move: str | int) -> int:
    """
    Convert a user move into its contract integer representation.

    rock     -> 0
    paper    -> 1
    scissors -> 2
    """
    if isinstance(move, int):
        if move not in INT_TO_MOVE:
            raise ValueError("Move must be 0, 1, or 2.")
        return move

    move_key = move.strip().lower()

    if move_key not in MOVE_TO_INT:
        raise ValueError("Move must be one of: rock, paper, scissors.")

    return MOVE_TO_INT[move_key]


def move_name(move: int) -> str:
    """
    Convert a contract move integer into a readable name.
    """
    if move not in INT_TO_MOVE:
        raise ValueError(f"Unknown move value: {move}")

    return INT_TO_MOVE[move]


def generate_salt() -> str:
    """
    Generate a random 32-byte salt as a 0x-prefixed hex string.
    """
    return "0x" + secrets.token_hex(32)


def make_commitment(
    *,
    contract_address: str,
    move: str | int,
    salt: str,
    player_address: str,
    game_id: int,
) -> bytes:
    """
    Compute the same commitment that Solidity computes:

    keccak256(abi.encodePacked(
        uint8 move,
        bytes32 salt,
        address player,
        uint256 gameId,
        address contract
    ))
    """
    move_int = normalize_move(move)

    return Web3.solidity_keccak(
        ["uint8", "bytes32", "address", "uint256", "address"],
        [
            move_int,
            salt,
            Web3.to_checksum_address(player_address),
            game_id,
            Web3.to_checksum_address(contract_address),
        ],
    )


def _secret_path(game_id: int, player_address: str) -> Path:
    safe_address = Web3.to_checksum_address(player_address)
    return LOCAL_SECRETS_DIR / f"game_{game_id}_{safe_address}.json"


def save_reveal_secret(
    *,
    game_id: int,
    player_address: str,
    move: str | int,
    salt: str,
    commitment: str,
) -> Path:
    """
    Save the local secret needed for reveal.

    This file must never be committed to Git.
    """
    LOCAL_SECRETS_DIR.mkdir(parents=True, exist_ok=True)

    move_int = normalize_move(move)
    path = _secret_path(game_id, player_address)

    data = {
        "game_id": game_id,
        "player_address": Web3.to_checksum_address(player_address),
        "move": move_int,
        "move_name": move_name(move_int),
        "salt": salt,
        "commitment": commitment,
    }

    path.write_text(json.dumps(data, indent=2))
    return path


def load_reveal_secret(*, game_id: int, player_address: str) -> dict:
    """
    Load the local move + salt needed for reveal.
    """
    path = _secret_path(game_id, player_address)

    if not path.exists():
        raise FileNotFoundError(
            f"No local reveal secret found for game {game_id} and player {player_address}.\n"
            "Without the original salt, this player cannot reveal their committed move."
        )

    return json.loads(path.read_text())


def delete_reveal_secret(*, game_id: int, player_address: str) -> None:
    """
    Delete local reveal secret after the game is finished.
    """
    path = _secret_path(game_id, player_address)

    if path.exists():
        path.unlink()

File: rps_frontend/src/utils/commit.ts
import { encodePacked, keccak256 } from "viem";
import type { Address, Hash } from "viem";

import { RPS_CONTRACT_ADDRESS } from "../config/contract";
import { asAddress } from "./contract";
import { devLog } from "./devLog";

const SECRET_KEY_PREFIX = "rps:reveal-secret";

const MOVE_TO_INT = {
  rock: 0,
  paper: 1,
  scissors: 2,
} as const;

const INT_TO_MOVE = {
  0: "rock",
  1: "paper",
  2: "scissors",
} as const;

export type MoveValue = 0 | 1 | 2;
export type MoveInput = MoveValue | string;

export type RevealSecret = {
  gameId: string;
  playerAddress: Address;
  move: MoveValue;
  moveName: string;
  salt: Hash;
  commitment: Hash;
};

type MakeCommitmentParams = {
  move: MoveInput;
  salt: Hash;
  playerAddress: string;
  gameId: bigint;
  contractAddress?: string;
};

function isMoveValue(move: number): move is MoveValue {
  return move === 0 || move === 1 || move === 2;
}

function secretKey(gameId: bigint | string, playerAddress: string): string {
  return `${SECRET_KEY_PREFIX}:${gameId.toString()}:${playerAddress.toLowerCase()}`;
}

export function normalizeMove(move: MoveInput): MoveValue {
  if (typeof move === "number") {
    if (isMoveValue(move)) return move;
    throw new Error("Move must be 0, 1, or 2.");
  }

  const key = move.trim().toLowerCase();

  if (key === "0" || key === "1" || key === "2") {
    return Number(key) as MoveValue;
  }

  if (key in MOVE_TO_INT) {
    return MOVE_TO_INT[key as keyof typeof MOVE_TO_INT];
  }

  throw new Error("Move must be rock, paper, or scissors.");
}

export function moveName(move: MoveInput): string {
  return INT_TO_MOVE[normalizeMove(move)];
}

export function generateSalt(): Hash {
  if (!globalThis.crypto?.getRandomValues) {
    throw new Error("Browser crypto is not available.");
  }

  const bytes = globalThis.crypto.getRandomValues(new Uint8Array(32));
  const hex = Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0"))
    .join("");

  return `0x${hex}` as Hash;
}

export function makeCommitment({
  contractAddress = RPS_CONTRACT_ADDRESS,
  move,
  salt,
  playerAddress,
  gameId,
}: MakeCommitmentParams): Hash {
  return keccak256(
    encodePacked(
      ["uint8", "bytes32", "address", "uint256", "address"],
      [
        normalizeMove(move),
        salt,
        asAddress(playerAddress),
        gameId,
        asAddress(contractAddress),
      ]
    )
  );
}

export function saveRevealSecret(secret: RevealSecret): void {
  devLog("commit:secret:save", {
    gameId: secret.gameId,
    playerAddress: secret.playerAddress,
    commitment: secret.commitment,
  });

  localStorage.setItem(
    secretKey(secret.gameId, secret.playerAddress),
    JSON.stringify(secret)
  );
}

export function loadRevealSecret(
  gameId: bigint,
  playerAddress: string
): RevealSecret | null {
  devLog("commit:secret:load", { gameId, playerAddress });
  const raw = localStorage.getItem(secretKey(gameId, playerAddress));

  if (!raw) return null;

  return JSON.parse(raw) as RevealSecret;
}

export function deleteRevealSecret(
  gameId: bigint,
  playerAddress: string
): void {
  devLog("commit:secret:delete", { gameId, playerAddress });
  localStorage.removeItem(secretKey(gameId, playerAddress));
}
